spring-kafka/src/main/java/org/springframework/kafka/core/DefaultTransactionIdSuffixStrategy.java

```diff
@@ -0,0 +1,109 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.kafka.core;
+
+import java.util.Map;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.springframework.lang.Nullable;
+import org.springframework.util.Assert;
+
+/**
+ * The {@link TransactionIdSuffixStrategy} implementation for managing transactional producer suffixes.
+ * If the {@link #maxCache} is greater than 0, the suffixes will be cached and reused.
+ * Otherwise, the suffixes will be generated on the fly.
+ *
+ * <p>
+ * Setting the {@link #setMaxCache(int)} enables caching to restrict the number of `transactional.id`.
+ * The cache is per `transactional.id` prefix. The cache size is limited by the {@link #maxCache}.
+ * Default is 0, which means no caching and restriction, so the `transactional.id` will be generated on the fly.
+ *
+ * @author Ilya Starchenko
+ *
+ * @since 3.2
+ */
+public class DefaultTransactionIdSuffixStrategy implements TransactionIdSuffixStrategy {
+
+	private final AtomicInteger transactionIdSuffix = new AtomicInteger();
+
+	private final Map<String, BlockingQueue<String>> suffixCache = new ConcurrentHashMap<>();
+
+	private int maxCache;
+
+	/**
+	 * Acquire the suffix for the transactional producer from the cache or generate a new one
+	 * if caching is disabled.
+	 * @param txIdPrefix the transaction id prefix.
+	 * @return the suffix.
+	 * @throws NoProducerAvailableException if caching is enabled and no suffixes are available.
+	 */
+	@Override
+	public String acquireSuffix(String txIdPrefix) {
+		Assert.notNull(txIdPrefix, "'txIdPrefix' must not be null");
+		BlockingQueue<String> cache = getSuffixCache(txIdPrefix);
+		if (cache == null) {
+			return String.valueOf(this.transactionIdSuffix.getAndIncrement());
+		}
+
+		String suffix = cache.poll();
+		if (suffix == null) {
+			throw new NoProducerAvailableException("No available transaction producer", txIdPrefix);
+		}
+		return suffix;
+	}
+
+	@Override
+	public void releaseSuffix(String txIdPrefix, String suffix) {
+		Assert.notNull(txIdPrefix, "'txIdPrefix' must not be null");
+		Assert.notNull(suffix, "'suffix' must not be null");
+		if (this.maxCache <= 0) {
+			return;
+		}
+		BlockingQueue<String> queue = getSuffixCache(txIdPrefix);
+		if (queue != null && !queue.contains(suffix)) {
+			queue.add(suffix);
+		}
+	}
+
+	/**
+	 * Set the maximum size for transaction producer cache.
+	 * @param maxCache the maxCache to set
+	 */
+	public void setMaxCache(int maxCache) {
+		Assert.isTrue(maxCache >= 0, "'maxCache' must greater than or equal 0");
```

Comment:
"'maxCache' must be greater than or equal to 0"?

Reply:
should we assert `maxCache > 0` to align with the adoc?

Reply:
I don't see any contradictions. The documentation states that if maxCache > 0, but the constraint specifies that it should not be negative. In this case, we could set maxCache to zero to disable caching.

Reply:
My first comment aims to point out the grammatical defect, my second one is about logic.

Reply:
> My first comment aims to point out the grammatical defect
Fixed.